### PR TITLE
Junction Snapshotter (POC)

### PR DESCRIPTION
**Pull request by RingerJK**
## Description

Junction Snapshotter (POC) 

- [ ] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Please describe the PR goals. Just the stuff needed to implement the fix / feature and a simple rationale. It could contain many check points if needed

### Implementation

Please include all the relevant things implemented and also rationale, clarifications / disclaimers etc. related to the approach used. It could be as self code companion comments

## Screenshots or Gifs

Please include all the media files to give some context about what's being implemented or fixed. It's not mandatory to upload screenshots or gifs, but for most of the cases it becomes really handy to get into the scope of the feature / bug being fixed and also it's REALLY useful for UI related PRs screenshot gif

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make core-update-api` (Core) / `$> make ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

### CODE REVIEW

**Comment by Guardiola31337** [comment]
A native crash was caught by the Robo tests when running `JunctionSnapshotActivity` https://console.firebase.google.com/u/0/project/mapbox-navigation-android/testlab/histories/bh.2e74b6b076d6fa53/matrices/5888843953099279755/executions/bs.1fa315d8ccf1ddc3/issues

```
08-17 05:50:12.434: W/Mbgl(14259): {Snapshotter}[OpenGL]: Not using Vertex Array Objects
08-17 05:50:12.435: W/Adreno-ES20(14259): <core_glRenderbufferStorage:659>: GL_INVALID_VALUE
08-17 05:50:12.437: A/libc(14259): /usr/local/google/buildbot/src/android/ndk-release-r21/external/libcxx/../../external/libcxxabi/src/abort_message.cpp:72: abort_message: assertion "terminating with uncaught exception of type std::runtime_error: Couldn't create framebuffer: incomplete attachment" failed
08-17 05:50:12.437: A/libc(14259): Fatal signal 6 (SIGABRT), code -6 in tid 16124 (Snapshotter)
```

File: examples/src/main/java/com/mapbox/navigation/examples/core/JunctionSnapshotActivity.kt [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+            else -> let {
+                Timber.w("Cannot find points to get bearing")
+                return
+            }
+        }
+
+        val bearing = TurfMeasurement.bearing(
+            points.first,
+            points.second
+        ).plus(180.0)
```
Comment (LukasPaczos) [inline examples/src/main/java/com/mapbox/navigation/examples/core/JunctionSnapshotActivity.kt:213]:
```suggestion
            geometryCoordinates.getOrNull(indexOfNearestPoint - 1) != null -> {
                geometryCoordinates[indexOfNearestPoint - 1] to geometryCoordinates[indexOfNearestPoint]
            }
            geometryCoordinates.getOrNull(indexOfNearestPoint + 1) != null -> {
                geometryCoordinates[indexOfNearestPoint] to geometryCoordinates[indexOfNearestPoint + 1]
            }
            else -> let {
                Timber.w("Cannot find points to get bearing")
                return
            }
        }

        val bearing = TurfMeasurement.bearing(
            points.first,
            points.second
        )
```

**Review by LukasPaczos — COMMENTED** [review]
Would be good to refactor to store fewer class fields. Also we're using internal APIs, but I think that's fine at a POC level.

I run into this crash when opening up the activity:
```
2020-08-20 12:53:21.589 14133-14133/com.mapbox.navigation.examples E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.mapbox.navigation.examples, PID: 14133
    java.lang.RuntimeException: Canvas: trying to draw too large(184443840bytes) bitmap.
        at android.graphics.RecordingCanvas.throwIfCannotDraw(RecordingCanvas.java:280)
        at android.graphics.BaseRecordingCanvas.drawBitmap(BaseRecordingCanvas.java:88)
        at android.graphics.drawable.BitmapDrawable.draw(BitmapDrawable.java:548)
        at android.widget.ImageView.onDraw(ImageView.java:1436)
        at android.view.View.draw(View.java:21975)
        at android.view.View.updateDisplayListIfDirty(View.java:20852)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at androidx.constraintlayout.widget.ConstraintLayout.dispatchDraw(ConstraintLayout.java:2023)
        at android.view.View.updateDisplayListIfDirty(View.java:20843)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at android.view.View.updateDisplayListIfDirty(View.java:20843)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at android.view.View.updateDisplayListIfDirty(View.java:20843)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at android.view.View.updateDisplayListIfDirty(View.java:20843)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at android.view.View.updateDisplayListIfDirty(View.java:20843)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at android.view.View.draw(View.java:21978)
        at com.android.internal.policy.DecorView.draw(DecorView.java:808)
        at android.view.View.updateDisplayListIfDirty(View.java:20852)
        at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:581)
        at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:587)
        at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:664)
        at android.view.ViewRootImpl.draw(ViewRootImpl.java:3767)
        at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:3495)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2779)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1745)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7768)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:967)
        at android.view.Choreographer.doCallbacks(Choreographer.java:791)
        at android.view.Choreographer.doFrame(Choreographer.java:726)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:952)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)

        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

File: examples/src/main/java/com/mapbox/navigation/examples/core/JunctionSnapshotActivity.kt (hunk 1 above)
Comment (RingerJK) [reply examples/src/main/java/com/mapbox/navigation/examples/core/JunctionSnapshotActivity.kt:217]: nice catch. fixed

**Comment by RingerJK** [comment]
@Guardiola31337 @LukasPaczos ready for another round

**Comment by RingerJK** [comment]
> Would be good to refactor to store fewer class fields.
suggestions?) 

> Also we're using internal APIs, but I think that's fine at a POC level.
fixed, avoid using internal API

**Comment by LukasPaczos** [comment]
I'm still hitting the same exception:
```
2020-09-03 15:43:22.644 4678-4678/com.mapbox.navigation.examples E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.mapbox.navigation.examples, PID: 4678
    java.lang.RuntimeException: Canvas: trying to draw too large(184443840bytes) bitmap.
        at android.graphics.RecordingCanvas.throwIfCannotDraw(RecordingCanvas.java:280)
        at android.graphics.BaseRecordingCanvas.drawBitmap(BaseRecordingCanvas.java:88)
        at android.graphics.drawable.BitmapDrawable.draw(BitmapDrawable.java:548)
        at android.widget.ImageView.onDraw(ImageView.java:1436)
        at android.view.View.draw(View.java:21975)
        at android.view.View.updateDisplayListIfDirty(View.java:20852)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at androidx.constraintlayout.widget.ConstraintLayout.dispatchDraw(ConstraintLayout.java:2023)
        at android.view.View.updateDisplayListIfDirty(View.java:20843)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at android.view.View.updateDisplayListIfDirty(View.java:20843)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at android.view.View.updateDisplayListIfDirty(View.java:20843)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at android.view.View.updateDisplayListIfDirty(View.java:20843)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at android.view.View.updateDisplayListIfDirty(View.java:20843)
        at android.view.View.draw(View.java:21707)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
        at android.view.View.draw(View.java:21978)
        at com.android.internal.policy.DecorView.draw(DecorView.java:808)
        at android.view.View.updateDisplayListIfDirty(View.java:20852)
        at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:581)
        at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:587)
        at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:664)
        at android.view.ViewRootImpl.draw(ViewRootImpl.java:3767)
        at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:3495)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2779)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1745)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7768)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:967)
        at android.view.Choreographer.doCallbacks(Choreographer.java:791)
        at android.view.Choreographer.doFrame(Choreographer.java:726)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:952)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

> suggestions?)

You could pass the variables as non-null function arguments instead of storing mutable objects as class fields. That might clean-up the code a lot I think.

**Comment by Guardiola31337** [comment]
I'm running into a native crash 💥 

```
09-03 09:46:01.700 7762-7967/com.mapbox.navigation.examples A/libc: /usr/local/google/buildbot/src/android/ndk-release-r21/external/libcxx/../../external/libcxxabi/src/abort_message.cpp:72: abort_message: assertion "terminating with uncaught exception of type std::runtime_error: Couldn't create framebuffer: incomplete attachment" failed
09-03 09:46:01.700 7762-7967/com.mapbox.navigation.examples A/libc: Fatal signal 6 (SIGABRT), code -6 in tid 7967 (Snapshotter)
09-03 09:46:01.801 198-198/? A/DEBUG: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
09-03 09:46:01.801 198-198/? A/DEBUG: Build fingerprint: 'google/hammerhead/hammerhead:6.0.1/M4B30Z/3437181:user/release-keys'
09-03 09:46:01.801 198-198/? A/DEBUG: Revision: '11'
09-03 09:46:01.801 198-198/? A/DEBUG: ABI: 'arm'
09-03 09:46:01.801 198-198/? A/DEBUG: pid: 7762, tid: 7967, name: Snapshotter  >>> com.mapbox.navigation.examples <<<
09-03 09:46:01.801 198-198/? A/DEBUG: signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
09-03 09:46:01.838 198-198/? A/DEBUG: Abort message: '/usr/local/google/buildbot/src/android/ndk-release-r21/external/libcxx/../../external/libcxxabi/src/abort_message.cpp:72: abort_message: assertion "terminating with uncaught exception of type std::runtime_error: Couldn't create framebuffer: incomplete attachment" failed'
09-03 09:46:01.838 198-198/? A/DEBUG:     r0 00000000  r1 00001f1f  r2 00000006  r3 9843f978
09-03 09:46:01.838 198-198/? A/DEBUG:     r4 9843f980  r5 9843f930  r6 0000000b  r7 0000010c
09-03 09:46:01.838 198-198/? A/DEBUG:     r8 9f6f987d  r9 9843f5a8  sl b6d65ec0  fp 000013f8
09-03 09:46:01.838 198-198/? A/DEBUG:     ip 00000006  sp 9843f088  lr b6d2bb61  pc b6d2df50  cpsr 400d0010
09-03 09:46:01.857 198-198/? A/DEBUG: backtrace:
09-03 09:46:01.858 198-198/? A/DEBUG:     #00 pc 00041f50  /system/lib/libc.so (tgkill+12)
09-03 09:46:01.858 198-198/? A/DEBUG:     #01 pc 0003fb5d  /system/lib/libc.so (pthread_kill+32)
09-03 09:46:01.858 198-198/? A/DEBUG:     #02 pc 0001c30f  /system/lib/libc.so (raise+10)
09-03 09:46:01.858 198-198/? A/DEBUG:     #03 pc 000194c1  /system/lib/libc.so (__libc_android_abort+34)
09-03 09:46:01.858 198-198/? A/DEBUG:     #04 pc 000174ac  /system/lib/libc.so (abort+4)
09-03 09:46:01.858 198-198/? A/DEBUG:     #05 pc 0001af23  /system/lib/libc.so (__libc_fatal+16)
09-03 09:46:01.858 198-198/? A/DEBUG:     #06 pc 00019549  /system/lib/libc.so (__assert2+20)
09-03 09:46:01.858 198-198/? A/DEBUG:     #07 pc 0027d92b  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.858 198-198/? A/DEBUG:     #08 pc 0027da2d  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.859 198-198/? A/DEBUG:     #09 pc 0027c17d  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.859 198-198/? A/DEBUG:     #10 pc 0027ba2b  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.859 198-198/? A/DEBUG:     #11 pc 0027b9f3  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so (__cxa_throw+74)
09-03 09:46:01.859 198-198/? A/DEBUG:     #12 pc 001844e3  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.859 198-198/? A/DEBUG:     #13 pc 00184377  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.859 198-198/? A/DEBUG:     #14 pc 0019216d  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.859 198-198/? A/DEBUG:     #15 pc 000f3157  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.859 198-198/? A/DEBUG:     #16 pc 000f28d9  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.859 198-198/? A/DEBUG:     #17 pc 00191f15  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.859 198-198/? A/DEBUG:     #18 pc 0019090f  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.860 198-198/? A/DEBUG:     #19 pc 00190cd3  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.860 198-198/? A/DEBUG:     #20 pc 000c4cb9  /data/app/com.mapbox.navigation.examples-1/lib/arm/libmapbox-gl.so
09-03 09:46:01.860 198-198/? A/DEBUG:     #21 pc 0003f45f  /system/lib/libc.so (_ZL15__pthread_startPv+30)
09-03 09:46:01.860 198-198/? A/DEBUG:     #22 pc 00019b43  /system/lib/libc.so (__start_thread+6)
09-03 09:46:01.852 198-198/? W/debuggerd: type=1400 audit(0.0:328500): avc: denied { read } for name="kgsl-3d0" dev="tmpfs" ino=6576 scontext=u:r:debuggerd:s0 tcontext=u:object_r:gpu_device:s0 tclass=chr_file permissive=0
09-03 09:46:03.182 202-812/? D/audio_hw_primary: disable_audio_route: reset and update mixer path: low-latency-playback
09-03 09:46:03.182 202-812/? D/audio_hw_primary: disable_snd_device: snd_device(2: speaker)
09-03 09:46:03.421 778-7992/? W/ActivityManager:   Force finishing activity com.mapbox.navigation.examples/.core.JunctionSnapshotActivity
09-03 09:46:03.423 198-198/? A/DEBUG: Tombstone written to: /data/tombstones/tombstone_02
09-03 09:46:03.423 198-198/? E/DEBUG: AM write failed: Broken pipe
```

Device: Nexus 5
Android version 6.0.1

File: examples/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -58,6 +58,9 @@
     <string name="title_guidance_view">GuidanceView</string>
     <string name="description_guidance_view">Demonstrates the use of a GuidanceView to show an instructional image at proper junctions.</string>
 
+    <string name="title_junction_snapshot_sample" translatable="false">Snapshotter Sample(POC)</string>
+    <string name="description_junction_snapshot_sample" translatable="false">Demonstrates junction view snapshot.</string>
```
Comment (Guardiola31337) [inline examples/src/main/res/values/strings.xml:62]: This shouldn't be `translatable = "false"`. Actually none of the resources here shouldn't 👀 https://github.com/mapbox/mapbox-navigation-android/pull/3438#issue-468106657 https://github.com/mapbox/mapbox-navigation-android/issues/2909 https://github.com/mapbox/mapbox-navigation-android/pull/3425#discussion_r471812462

**Comment by RingerJK** [comment]
@LukasPaczos could you run it once again?

**Comment by RingerJK** [comment]
@LukasPaczos @Guardiola31337 ready for the next round

**Comment by Guardiola31337** [comment]
This is working now 🎉 

junction_snapshotter_poc_nexus_5

We should probably place the route line behind the shields so they're not "hidden". That's the default behavior in the UI SDK and upcoming maneuver arrows on top. Are we going to include arrows as part of the POC? 🤔

**Review by LukasPaczos — APPROVED** [review]

**Comment by RingerJK** [comment]
@Guardiola31337 moved route line under shields 👍

**Comment by RingerJK** [comment]
> @Guardiola31337 moved route line under shields 👍 I thought that arrows weren't a part of POC, correct me if I'm wrong

@asinghal22 should we include arrows to Snapshotter POC?

**Comment by asinghal22** [comment]
@RingerJK Since this is a PoC only conduct tasks that require us to validate that the feature is feasible from engg. perspective ie. if adding arrows is trivial then don't bother. If they require engineering deliberation / architecture then please do add that.

BTW: I don't think that tilt of 60 degrees cuts it. We need to be in a 'horizon' view. I will do some research on what tilt is that but we might need that to be more than 60.

**Comment by RingerJK** [comment]
Let's merge it for now. Arrows might be added later by requirements.

**Pull request merged**